File: 77-combinations/combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let result = [];
    let arr = [];

    solve(result, arr, 1, n, k);

    return result;
};

function solve(result, arr, i, n, k) {
    
    if(arr.length === k) {
        result.push([...arr]);
        return;
    }

    if(i > n){
        return;
    }
    
    arr.push(i);
    solve(result, arr, i+1, n, k);
    arr.pop();

    solve(result, arr, i+1, n, k);
};

File: 77-combinations/combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> list;
        vector<vector<int>> ans;

        solve(1, list, ans, n, k);
        return ans;
    }
    void solve(int i, vector<int> &list, vector<vector<int>> &ans, int n, int k)
    {
        if(list.size() == k)
        {
            ans.push_back(list);
            return;
        }

        for(int it = i; it <= n; it++)
        {
            list.push_back(it);

            solve(it+1, list, ans, n, k);

            list.pop_back();
        }
    }
};